src/arch/host/include/arch/string.h

```diff
@@ -32,8 +32,11 @@ static inline int arch_memcpy_s(void *dest, size_t dest_size,
 	if (!dest || !src)
 		return -EINVAL;
 
-	if ((dest + dest_size >= src && dest + dest_size <= src + src_size) ||
-		(src + src_size >= dest && src + src_size <= dest + dest_size))
+	if (dest == src)
+		return -EINVAL;
+
+	if ((dest + dest_size > src && dest + dest_size < src + src_size) ||
```

Comment:
It's still invalid f.e. it should fail for that case:
```c
char a[8] = {0};
assert_int_not_equal(memcpy_s(a, 2, a + 1, 1), 0);
```
Also there are unnecessary additions in current implementation.
I think it's the correct solution:
```c
if ((dest >= src && dest < (src + src_size)) ||
		(src >= dest && src < (dest + dest_size)))
```

Reply:
Also with code above check for dst==src is not needed, however with 0-sized arrays it will pass this check, but it shouldn't do anything with that size anyway

Reply:
> however with 0-sized arrays it will pass this check

Because this is the bare minimum we should know and check ourselves.

> It's still invalid f.e. it should fail for that case:
> 
> ```c
> char a[8] = {0};
> assert_int_not_equal(memcpy_s(a, 2, a + 1, 1), 0);
> ```

> Please use @jajanusz's solution.

I don't think it should fail in this case. You say "array `a` has two bytes, please copy one byte from address `a+1` to the address `a`" Sounds perfectly legal to me, don't you think?
What your memcpy_s do is in essence a[0] = a[1] -> Why would we forbid such copy?

Reply:
> Sounds perfectly legal to me, don't you think?

We may think whatever we want, but standard says clearly that buffers in memcpy **cannot** overlap. These arguments are invalid. So yea it's an error.

Reply:
@jajanusz but they don't overlap in your example, do they? Also what exactly do you read? I guess you read memcpy specification but here we discuss memcpy_s, a wrapper function.

Reply:
> but they don't overlap in your example, do they?

last byte of dst is first byte of src, they overlap on tha 1 byte





> Also what exactly do you read? I guess you read memcpy specification but here we discuss memcpy_s, a wrapper function.

It's invalid for both. For `memcpy` it results with *undefined behavior*, for `memcpy_s` it's an error.